File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 042583. 다리를 지나는 트럭/kotlin/Solution.kt
class Solution {
    fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
        val bridge = Bridge(bridge_length, weight)

        for (truck in truck_weights) {
            while (!bridge.dequeueAndEnqueue(truck));
        }
        while (!bridge.isEmpty()) bridge.dequeueAndEnqueueNothing()
        return bridge.elapsedTimeFromSetUp
    }
}

class Bridge(
    cap: Int,
    private val maxWeight: Int,
) {

    var elapsedTimeFromSetUp = 0
    private var c = cap + 1
    private val arr = IntArray(c + 1)
    private var front = 0
    private var rear = cap
    private var weight = 0

    /**
     * 다리에서, 제일 앞에 있던 요소를 다리 밖으로 빼고, 새로운 요소를 다리에 삽입 시도합니다.
     * 다리에 삽입하는 것에 성공하면 true를 반환합니다.
     * 다리에 삽입하는 것에 실패하면 false를 반환합니다.
     * 이 과정에서 시간이 경과합니다.
     */
    fun dequeueAndEnqueue(car: Int): Boolean {
        elapsedTimeFromSetUp ++

        if (weight - arr[(front + 1 + c) % c] + car <= maxWeight) {
            dequeue()
            enqueue(car)
            return true
        } else {
            dequeue()
            enqueue(0)
            return false
        }
    }

    /**
     * 다리에서 제일 앞의 요소를 다리 밖으로 빼고 아무 것도 통과시키지 않습니다.
     * 이 과정에서 시간이 경과합니다.
     */
    fun dequeueAndEnqueueNothing() {
        dequeueAndEnqueue(0)
    }

    /**
     * 다리에 아무 트럭도 없으면 true를 반환합니다.
     */
    fun isEmpty() = weight == 0

    private fun enqueue(e: Int) {
        arr[((rear + 1 + c) % c).also { rear = it }] = e
        weight += e
    }

    private fun dequeue(): Int = arr[((front + 1 + c) % c).also { front = it; weight -= arr[front] }]
}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 042583. 다리를 지나는 트럭/python/solution.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0 # 현재 시각
    bridge_queue = deque([0] * bridge_length) # 다리 위를 지나가는 차량의 상태를 큐로 표현
    current_bridge_weight = 0 # 다리 위에 놓인 무게 합
    truck_cursor = 0 # 통과해야할 트럭의 번호

    while True:
        time += 1
        first = bridge_queue.popleft() # 다리 맨 앞의 요소를 제거
        current_bridge_weight -= first

        # 트럭을 실었을 때 다리가 견딜 수 있으면
        if truck_cursor < len(truck_weights) and current_bridge_weight + truck_weights[truck_cursor] <= weight:
            current_bridge_weight += truck_weights[truck_cursor]
            bridge_queue.append(truck_weights[truck_cursor])
            truck_cursor += 1

        # 다리에 0만 삽입(공기만 지나감)
        else:
            bridge_queue.append(0)

        # 다리 위의 무게가 0이 됐을 때 (모든 트럭이 지나갔을 때) 종료
        if current_bridge_weight == 0:
            break
    return time
